File: patches/exploding_villagers.cpp
#include "exploding_villagers.h"
#include <set>
#include "genie/dat/DatFile.h"
#include "ids.h"


void configureExplodingVillagers(genie::DatFile *df) {
    std::set<int> villagers = {
        ID_FISHING_SHIP,        ID_TRADE_COG,           ID_TRADE_CART_EMPTY,   ID_TRADE_CART_FULL,
        ID_VILLAGER_BASE_M,     ID_VILLAGER_BASE_F,     ID_VILLAGER_FARMER_M,  ID_VILLAGER_FARMER_F,
        ID_VILLAGER_SHEPHERD_M, ID_VILLAGER_SHEPHERD_F, ID_VILLAGER_FORAGER_M, ID_VILLAGER_FORAGER_F,
        ID_VILLAGER_HUNTER_M,   ID_VILLAGER_HUNTER_F,   ID_VILLAGER_FISHER_M,  ID_VILLAGER_FISHER_F,
        ID_VILLAGER_WOOD_M,     ID_VILLAGER_WOOD_F,     ID_VILLAGER_GOLD_M,    ID_VILLAGER_GOLD_F,
        ID_VILLAGER_STONE_M,    ID_VILLAGER_STONE_F,    ID_VILLAGER_BUILDER_M, ID_VILLAGER_BUILDER_F,
        ID_VILLAGER_REPAIRER_M, ID_VILLAGER_REPAIRER_F,
    };

    for (genie::Civ &civ : df->Civs) {
        for (int villager_id : villagers) {
            civ.Units.at(villager_id).DeadUnitID = ID_SABOTEUR;
            std::cout << "Patched Villager unit " << villager_id << " for civ " << civ.Name << "\n";
        }
        genie::Unit &saboteur = civ.Units.at(ID_SABOTEUR);
        saboteur.HitPoints = 0;
        saboteur.Type50.Attacks.at(0).Amount = 50;
        saboteur.Type50.Attacks.at(1).Amount = 90;
        saboteur.Type50.Attacks.at(2).Amount = 0;
        saboteur.Type50.MaxRange = 2;
        saboteur.Type50.BlastAttackLevel = 1;  // cut trees
        saboteur.TrainSound = -1;
        saboteur.WwiseTrainSoundID = 0;  // prevent melee unit train sound from playing
        std::cout << "Patched Saboteur unit for civ " << civ.Name << "\n";
    }
}

File: mods/exploding_villagers.py
import logging

from genieutils.civ import Civ
from genieutils.datfile import DatFile
from genieutils.unit import Unit

from mods.ids import SABOTEUR, CLASS_TRADE_BOAT, CLASS_CIVILIAN, CLASS_TRADE_CART, CLASS_FISHING_BOAT
from mods.util import clone

NAME = 'exploding-villagers'
NAME_2 = 'exploding-villagers-extreme'

EXPLODING_CLASSES = [
    CLASS_TRADE_BOAT,
    CLASS_CIVILIAN,
    CLASS_TRADE_CART,
    CLASS_FISHING_BOAT,
]


def clone_saboteur(civ: Civ, version: str) -> int:
    clone_unit_id = len(civ.units)
    cloned_unit = clone(civ.units[SABOTEUR], version)
    cloned_unit.id = clone_unit_id
    cloned_unit.hit_points = -1
    cloned_unit.type_50.max_range = 2
    cloned_unit.type_50.blast_attack_level = 1
    cloned_unit.train_sound = -1
    cloned_unit.wwise_train_sound_id = 0
    logging.info(f'Cloned unit {clone_unit_id} ({cloned_unit.name}) for civ {civ.name}')
    civ.units.append(cloned_unit)
    return clone_unit_id


def should_explode(unit: Unit) -> bool:
    return unit.class_ in EXPLODING_CLASSES if unit else False


def patch_villager_units(data: DatFile, nerf_saboteur: bool):
    for civ in data.civs:
        saboteur_id = clone_saboteur(civ, data.version)
        if nerf_saboteur:
            type_50 = civ.units[saboteur_id].type_50
            type_50.attacks[0].amount = 50
            type_50.attacks[1].amount = 90
            type_50.attacks[2].amount = 0
        for unit in civ.units:
            if should_explode(unit):
                unit.dead_unit_id = saboteur_id
                logging.info(f'Patched unit with id {unit.id} ({unit.name}) for civ {civ.name}')


def mod(data: DatFile):
    patch_villager_units(data, nerf_saboteur=True)


def mod_2(data: DatFile):
    patch_villager_units(data, nerf_saboteur=False)
